### PR TITLE
LC-9 connection request

### PR DESCRIPTION
**Pull request by bambooSocks**
I know it is not done yet ... two tests are not properly written, but I was wondering whether you could have a look at it now so I can see if I am doing the stuff correctly

### CODE REVIEW

File: src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        String message;
+    }
+
+    /**
+     * JSON Mapping class for JSONs involved with creating a new Connection Request
+     */
+    @Value
+    public static class NewConnectionRequestDTO {
+        @NotNull
+        Long senderId;
```
Comment (Krismix1) [inline src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java:39]: We can/should also check for minimum value using something like `Min(1)`.

File: src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    public Iterable<ConnectionRequest> getAllConnectionRequest() {
+        return connectionRequestService.getAll();
+    }
+
+    @DeleteMapping("/requests/{id}")
+    public ResponseEntity<?> deleteConnectionRequest(@PathVariable long id) {
+        try {
+            connectionRequestService.delete(id);
+        } catch (NoSuchElementException e) {
+            return ResponseEntity.notFound().build();
```
Comment (Krismix1) [inline src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java]: I think typically the API should also return a message saying something like `Resource with id %d not found` (or maybe without the ID). This way the consumer of the API can differentiate between 404 meaning that the whole URL was not found or if just the resource with that ID.

Another way is to create a `ResourceNotFound` exception which will be raised in such scenarios. Then the global exception handler would catch it and return a 404 with the message. This could potentially keep the code more DRY.

File: src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        }
+        return ResponseEntity.noContent().build();
+    }
+
+    @PostMapping("/requests/accept/{id}")
+    public ResponseEntity<?> acceptConnectionRequest(@PathVariable long id) {
+        try {
+            connectionRequestService.accept(id);
+        } catch (NoSuchElementException e) {
+            return ResponseEntity.notFound().build();
```
Comment (Krismix1) [inline src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java]: The same here with the 404.

File: src/main/java/org/autopotato/linkedinclonebackend/repositories/CrudRepository.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
+/*
```
Comment (Krismix1) [inline src/main/java/org/autopotato/linkedinclonebackend/repositories/CrudRepository.java]: Let's remove the license comment.

File: src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java (hunk 1 above)
Comment (bambooSocks) [reply src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java:39]: I have added the annotations

File: src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java (hunk 2 above)
Comment (bambooSocks) [reply src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java]: I switched to `ResourceNotFoundException`

File: src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java (hunk 3 above)
Comment (bambooSocks) [reply src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java]: same here

File: src/main/java/org/autopotato/linkedinclonebackend/repositories/CrudRepository.java (hunk 4 above)
Comment (bambooSocks) [reply src/main/java/org/autopotato/linkedinclonebackend/repositories/CrudRepository.java]: done

File: src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+     * Deny or cancel a {@link ConnectionRequest}
+     * @param id ID of the {@link ConnectionRequest} to be deleted
+     * @return Not Found if {@param id} is not in the system
+     *         No Content otherwise
+     */
+    @DeleteMapping("/requests/{id}")
+    public ResponseEntity<?> deleteConnectionRequest(@PathVariable long id) {
+        try {
+            connectionRequestService.delete(id);
+        } catch (ResourceNotFoundException e) {
```
Comment (Krismix1) [inline src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java]: Let's handle the exception in `ControllerExceptionHandler` as this will be a widely used exception.

File: src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+            return ResponseEntity.status(HttpStatus.NOT_FOUND).body(e.getMessage());
```
Comment (Krismix1) [inline src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java]: The response body should contain a JSON object, not just a string.

File: src/test/java/org/autopotato/linkedinclonebackend/unit/controllers/ConnectionControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+    @BeforeEach
+    final void setUp() {
+        jsonMapper = new ObjectMapper();
+    }
+
+    @Test
+    final void createConnectionRequest() {
+        newConnectionRequestDTO =
+            new ConnectionController.NewConnectionRequestDTO(1L, 2L);
```
Comment (Krismix1) [inline src/test/java/org/autopotato/linkedinclonebackend/unit/controllers/ConnectionControllerTest.java]: Do you need to explicitly say it's a long?

File: src/test/java/org/autopotato/linkedinclonebackend/unit/services/ConnectionRequestServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+    @Mock
+    ConnectionService connectionService;
+
+    @Mock
+    MockConnectionRequestRepository mockConnectionRequestRepository;
+
+    @BeforeEach
+    void setUp() {
+        MockitoAnnotations.initMocks(this);
```
Comment (Krismix1) [inline src/test/java/org/autopotato/linkedinclonebackend/unit/services/ConnectionRequestServiceTest.java]: What if we annotate the whole test class with `@ExtendWith(MockitoExtension.class)` instead of this?

File: src/main/java/org/autopotato/linkedinclonebackend/controllers/ResourceNotFoundException.java [hunk 9]
```diff
@@ -0,0 +1,15 @@
+package org.autopotato.linkedinclonebackend.controllers;
+
+import lombok.EqualsAndHashCode;
+import lombok.Value;
+
+@EqualsAndHashCode(callSuper = true)
+@Value
+public class ResourceNotFoundException extends Exception {
```
Comment (Krismix1) [inline src/main/java/org/autopotato/linkedinclonebackend/controllers/ResourceNotFoundException.java]: Usually, I would use the `RuntimeException` as the base class. But I can't say that there is a straightforward answer.
I guess we can try using caught exceptions and see how it goes.

For inspiration:
https://stackoverflow.com/questions/27578/when-to-choose-checked-and-unchecked-exceptions
https://stackoverflow.com/questions/499437/in-java-when-should-i-create-a-checked-exception-and-when-should-it-be-a-runti
https://stackoverflow.com/questions/6115896/understanding-checked-vs-unchecked-exceptions-in-java

File: src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java (hunk 5 above)
Comment (bambooSocks) [reply src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java]: ok I moved it

File: src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java (hunk 6 above)
Comment (bambooSocks) [reply src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java]: I used the ExceptionResponseDTO

File: src/test/java/org/autopotato/linkedinclonebackend/unit/controllers/ConnectionControllerTest.java (hunk 7 above)
Comment (bambooSocks) [reply src/test/java/org/autopotato/linkedinclonebackend/unit/controllers/ConnectionControllerTest.java]: if you mean the literal then yes ... it complains that it is an integer since it takes `Long` not `long` ... no implicit conversion there from `int` :/

File: src/test/java/org/autopotato/linkedinclonebackend/unit/services/ConnectionRequestServiceTest.java (hunk 8 above)
Comment (bambooSocks) [reply src/test/java/org/autopotato/linkedinclonebackend/unit/services/ConnectionRequestServiceTest.java]: I don't see a difference tho ... I was following tutorial and it worked ... but I guess for consistency I can change it

File: src/main/java/org/autopotato/linkedinclonebackend/controllers/ResourceNotFoundException.java (hunk 9 above)
Comment (bambooSocks) [reply src/main/java/org/autopotato/linkedinclonebackend/controllers/ResourceNotFoundException.java]: I agree, I changed it to `RuntimeException` ... I always forget that it exists 🙄

File: src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+
+    /**
+     * Deny or cancel a {@link ConnectionRequest}
+     * @param id ID of the {@link ConnectionRequest} to be deleted
+     * @return Not Found if {@param id} is not in the system
+     *         No Content otherwise
+     */
+    @DeleteMapping("/requests/{id}")
+    public ResponseEntity<?> deleteConnectionRequest(@PathVariable long id)
+        throws ResourceNotFoundException {
```
Comment (Krismix1) [inline src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java]: We can probably remove the `throws ResourceNotFoundException` part(s) (also from tests). Is there a reason to keep them?

File: src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java (hunk 1 above)
Comment (Krismix1) [reply src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java:39]: We could also write tests for negative values and for 0.

File: src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java (hunk 10 above)
Comment (bambooSocks) [reply src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java]: ok I removed it

File: src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java (hunk 1 above)
Comment (bambooSocks) [reply src/main/java/org/autopotato/linkedinclonebackend/controllers/ConnectionController.java:39]: I already did

File: src/test/java/org/autopotato/linkedinclonebackend/unit/controllers/ConnectionControllerTest.java [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+    ConnectionRequestService connectionRequestService;
+
+    NewConnectionRequestDTO newConnectionRequestDTO;
+
+    @Autowired
+    MockMvc mockMvc;
+
+    private ObjectMapper jsonMapper;
+
+    public final NewConnectionRequestDTO[] conenctionOptions() {
```
Comment (Krismix1) [inline src/test/java/org/autopotato/linkedinclonebackend/unit/controllers/ConnectionControllerTest.java]: There is a typo here in `conenctionOptions`. Also, I fail to find where this is used.

Comment (bambooSocks) [reply src/test/java/org/autopotato/linkedinclonebackend/unit/controllers/ConnectionControllerTest.java]: oh, I will handle that.. stand by

Comment (bambooSocks) [reply src/test/java/org/autopotato/linkedinclonebackend/unit/controllers/ConnectionControllerTest.java]: done

**Review by Krismix1 — APPROVED** [review]
Great work!
